doc/guild.1.in
.\" The macros used below are those described in groff_man(7) and
 \" groff_man_style(7).
.
.TH guild 1 "2025-12-12" "@GUILE_VERSION@" "GNU Guile"
.
.SH NAME
guild, guile-tools \- run Guile-related commands
.
.SH SYNOPSIS
.SY guild
.B compile
.YS
.SY guild
.B disassemble
.YS
.SY guild
.B display\-commentary
.YS
.SY guild
.B doc\-snarf
.YS
.SY guild
.B help
.YS
.SY guild
.B use2dot
.YS
.SY .\|.\|.
.YS
.SH DESCRIPTION
Run Guile-related commands; \(dqguild help --all\(dq will provide a
complete list.
.B guile\-tools
is an older, deprecated name for this program.
.
.SH SUBCOMMANDS
.TP
.B help \fR[\fB--all \fR| \fICOMMAND]
Display a list of
.B \-\-all
available subcommands, information about
.IR COMMAND ,
or summary information.
.TP
.B compile \fR[\fIARG \fR\&.\|.\|.\&]
Invoke the Guile compiler.
.TP
.B disassemble \fR[\fIFILE \fR\&.\|.\|.\&]
Disassemble a compiled .go file.
.TP
.B display\-commentary \fR[\fIREF \fR\&.\|.\|.\&]
Display Commentary section from
.IR REF1 , \ REF2
and so on.  Each
.I REF
may be a filename or module name (list of symbols).  In the latter case,
a filename is computed by searching
.BR %load-path.
.TP
.BI doc\-snarf \ FILE
Read a Scheme source file to extract docstrings in the format specified
in
.B DOCSTRINGS
below, and infer a procedure prototype from procedure definition lines
starting with
.BR "(define... )" .
.
Currently, two output modes are implemented: texinfo and plaintext.
Default is plaintext, texinfo can be switched on with the
.B \-\-texinfo
or
.B \-t
command line option.
.TP
.BI use2dot \ \fR[\fB\-m \ \fIMODULE] \ [\fIFILE \fR\&.\|.\|.\&]
Display to stdout a graphviz DOT specification that describes module
dependencies in the
.IR FILE s.

A top-level
.B use-modules
form or a
.BR :use-module
component in a
.B define-module
form results in a \(dqsolid\(dq style edge.

An
.B :autoload
component in a
.B define-module
form results in a \(dqdotted\(dq style edge with label \(dqN\(dq
indicating that N names are responsible for triggering the autoload.
(The \(dqN\(dq label is not implemented.)

A top-level
.B load or
.B primitive-load form results in a a \(dqbold\(dq style
edge to a node named with either the file name if the
.B load
argument is a string, or \(dq[computed in \fIFILE\fR]\(dq otherwise.

The
.BR \-\-default-module
.I MODULE
(or
.BR \-m)
option sets
.I MODULE
as the default for top-level
.B use-modules
forms that do not follow a
.B define-module
form in a file.
.I MODULE
should be be a list, or
.BR #f ,
in which case such top-level
.B use-modules forms are effectively
ignored.  The default value is
.B (guile-user).
.
.SH DOCSTRINGS
A docstring can span multiple lines and begins with a line that starts
with \(dq;;\(dq (two semicolons and a space). A docstring ends with a
line that begins with \(dq(define ...)\(dq or with one or more
\(dqoption\(dq lines that begin with \(dq;;-\(dq (two semicolons and a
dash), which is then followed by a keyword, a colon, and a string.

Additionally, \(dqstandard internal docstrings\(dq (for Scheme source)
are recognized and output as \(dqoptions\(dq.  The output formatting is
likely to change in the future.

Here's an example:

.RS 4
.EX
;; This procedure foos, or bars, depending on the argument @var{braz}.
;;-Author: Martin Grabmueller
(define (foo/bar braz)
  (if braz 'foo 'bar))
.EE
.RE

which produces this when the output format is texinfo:

.RS 4
.EX
#!
.
foo/bar
@deffn procedure foo/bar braz
This procedure foos, or bars, depending on the argument @var{braz}.
@c Author: Martin Grabmueller
@end deffn
!#

.EE
.RE

or this when the output format is plain text:

.RS 4
.EX
#!
Procedure: foo/bar braz
This procedure foos, or bars, depending on the argument @var{braz}.
 Author: Martin Grabmueller
^L
!#
.EE
.RE
.
.SH SEE ALSO
.
.IR dot (1),
.IR guile (1),
and for much more detail, the
.UR https://www.gnu.org/software/guile/manual/html_node/index.html
Guile Reference Manual
.UE
which may also be available via
.BR info\ guile
or
.B M\-x info
in emacs.


NEWS
Guile NEWS --- history of user-visible changes.
Copyright (C) 1996-2025 Free Software Foundation, Inc.
See the end for copying conditions.

Please send Guile bug reports to bug-guile@gnu.org.


Changes in 3.0.11 (since 3.0.10)

* Notable changes

** Guile supports SRFI-207: String notated bytevectors (bytestrings)

Guile can now read and write #u8"\xe2;\x88;\x9e; Improbability\n" style
bytestrings when they are enabled by `(read-enable 'bytestrings)` and
`(print-enable 'bytestrings)` respectively. See "SRFI-207 String-notated
bytevectors" in the info pages for additional information.

All of the SRFI-207 procedures have also been added, including support
for hex and base64 encoding and decoding.

** R7RS `cond-expand` now supports `else` (<https://debbugs.gnu.org/71304>)

** The SRFI-64 module for test suites has been rewritten

A new, backward-compatible implementation of (srfi srfi-64) has been
contributed by Tomas Volf.  The reference implementation, which was used
so far, had a number of issues on Guile that made some of its interface
downright unusable (e.g., <https://bugs.gnu.org/72378>), non-conforming
(e.g., <https://bugs.gnu.org/72383>), or buggy (e.g.,
<https://bugs.gnu.org/72372>).

* New interfaces and functionality

** Guile now supports SRFI-197 (Pipeline Operators)

Guile now provides chain, chain-and, chain-when, chain-lambda, nest, and
nest-reverse.

** Guile now supports SRFI-244 (Multiple-value Definitions)

Guile now provides syntax 'define-values' under the module (srfi
srfi-244). This syntax existed for a long time in the core of Guile,
now it is also exported via srfi-244 to help writing portable code.

** PEG parser

PEG grammar parser in (ice-9 peg string-peg) has been rewritten to cover
all the functionality defined in <https://bford.info/pub/lang/peg.pdf>.

The 'not-in-range' pattern was also added to (ice-9 peg); it is
available from PEG strings via '[^...]'.

** GOOPS: Introduce new forms method* and define-method*

The module (oop goops) now exports method* and define-method* which are
the counterparts of lambda* and define* in they provide support for
advanced argument handling such as optional and keyword arguments. The
implementation fully supports (next-method) calls, also for keyword
arguments. The new syntax is documented in the Guile Reference manual.

** (ice-9 rdelim) has new line and field iteration procedures

(ice-9 rdelim) has three new procedures: for-rdelim-from-port,
for-delimited-from-port and for-line-in-file. Of these, for-line-in-file
is helpful in the common situation where you want a procedure applied to
every line in a file.

** Add procedure (%inhibit-welcome-message) to silence the REPL welcome

If ((@ (system repl repl) %inhibit-welcome-message) #t) is set in
$HOME/.guile, the welcome message in the REPL is suppressed.

** Improve 'load-foreign-library' handling of DLLs

The non-libltdl load-foreign-library introduced in 3.0.6 does not handle
some common cases with libtool-generated DLLs.  It has been updated
to search for DLLs that have a version number appended to the name
by libtool, such as libfoo-1.dll.

Also, it has been updated to do library renaming for MSYS.  On Cygwin,
when the #:rename-on-cygwin? option is #t, it already had the capability
to search for "libfoo" as "cygfoo.dll".  It has been updated to add the
capability to search for "libfoo" as "msys-foo.dll" on MSYS.

The load-foreign-library option #:rename-on-cygwin? has been changed to
#:host-type-rename?, and handles both Cygwin and MSYS.

** Make piped-process and system* available on systems without fork

Now that piped-process and system* are implemented in terms of
`posix_spawn', they can be made available on systems without fork().
Note that currently Guile does use fork in piped-process to set exit
codes, so piped-process on systems without fork will have a different
behavior with regards to exit codes.

** MinGW: remove kill, getaffinity, setaffinity, getpriority, and setpriority

Now that Guile uses a posix_spawn module, these Win32 emulations of
POSIX functions no longer work.  They relied on Guile keeping an
internal handle-to-pid table, which is not part of the new posix_spawn
code.  They are removed on MinGW.

** MinGW: waitpid returns ENOSYS if options are non-zero

When Guile switched to using posix_spawn module, Guile's Win32 emulation
of waitpid could no longer emulate ENOHANG.  It relied on Guile keeping
an internal handle-to-pid table.  It now returns ENOSYS on any non-zero
waitpid option.

** The JIT (lightening) supports the riscv architecture

* Performance improvements

** 'copy-file' now relies on 'sendfile' rather than a read/write loop

** The performance of 'letrec' compilation has been improved

* Miscellaneous improvements

** Some array function error messages have been improved

** The psyntax implementation has been improved

** Many Cygwin/MSYS incompatibilities have been fixed

* Documentation updates

** The '-C', '--rNrs', and '--language' arguments are now included in 'guile(1)'

** Now documented: 'peek', 'pk', 'in-vicinity', and 'use-modules' '#:hide'

** The way to avoid SXML escaping has been documented

** SRFI-64 (test suite) documentation has been added

** 'setlocale' now mentions that it raises a 'system-error' for missing locales

* Changes to the distribution

* Build system changes

** -Werror=array-bounds is now added to CFLAGS when available

This catches <https://bugs.gnu.org/76907>

** Parallel builds should no longer fail due to a missing guile-procedures.txt

** Some sparse file tests are now skipped on Darwin hosts

While APFS does support sparse files, they behave differently.

** Guile now compiles with GCC 14 and musl on 32-bit systems

** Compilation with '-Werror=undef' works again (broken in 3.0.10)

** Some missing, readonly, and typical 'copy-file' tests have been added

** The test suite now supports SRFI-64 based tests

This support should be considered internal for now and subject to
change.


module/guild.scm
;;;; guild.scm --- running scripts bundled with Guile
;;;; Andy Wingo <wingo@pobox.com> --- April 2009
;;;; 
;;;; 	Copyright (C) 2009, 2010, 2011, 2013, 2014 Free Software Foundation, Inc.
;;;; 
;;;; This library is free software; you can redistribute it and/or
;;;; modify it under the terms of the GNU Lesser General Public
;;;; License as published by the Free Software Foundation; either
;;;; version 3 of the License, or (at your option) any later version.
;;;; 
;;;; This library is distributed in the hope that it will be useful,
;;;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
;;;; Lesser General Public License for more details.
;;;; 
;;;; You should have received a copy of the GNU Lesser General Public
;;;; License along with this library; if not, write to the Free
;;;; Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
;;;; Boston, MA 02110-1301 USA

(define-module (guild)
  #:use-module (ice-9 getopt-long)
  #:use-module (ice-9 command-line)
  #:autoload (ice-9 format) (format))

;; Hack to provide scripts with the bug-report address.
(module-define! the-scm-module
                '%guile-bug-report-address
                "bug-guile@gnu.org")


(define *option-grammar*
  '((help (single-char #\h))
    (version (single-char #\v))))

(define (display-version)
  (version-etc "GNU Guile"
               (version)
               #:command-name "guild"
               #:license *LGPLv3+*))

(define (find-script s)
  (resolve-module (list 'scripts (string->symbol s)) #:ensure #f))

(define (main args)
  (if (defined? 'setlocale)
      (catch 'system-error
        (lambda ()
          (setlocale LC_ALL ""))
        (lambda args
          (format (current-error-port)
                  "warning: failed to install locale: ~a~%"
                  (strerror (system-error-errno args))))))

  (let* ((options (getopt-long args *option-grammar*
                               #:stop-at-first-non-option #t))
         (args (option-ref options '() '())))
    (cond
     ((option-ref options 'help #f)
      (apply (module-ref (resolve-module '(scripts help)) 'main) args)
      (exit 0))
     ((option-ref options 'version #f)
      (display-version)
      (exit 0))
     ((find-script (if (null? args) "help" (car args)))
      => (lambda (mod)
           (exit (apply (module-ref mod 'main) (if (null? args)
                                                   '()
                                                   (cdr args))))))
     (else
      (format (current-error-port)
              "guild: unknown script ~s~%" (car args))
      (format (current-error-port)
              "Try `guild help' for more information.~%")
      (exit 1)))))
